modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,346 @@
+			 * Interpret "lazy" as not in viewport. Any other value can be
+			 * interpreted as in viewport (realistically only "eager" or `false`
+			 * to force-omit the attribute are other potential values).
+			 */
+			if ( 'lazy' === $attr['loading'] ) {
+				$maybe_in_viewport = false;
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		// Logic to handle a `fetchpriority` attribute that is already provided.
+		if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+			/*
+			 * If the image was already determined to not be in the viewport (e.g.
+			 * from an already provided `loading` attribute), trigger a warning.
+			 * Otherwise, the value can be interpreted as in viewport, since only
+			 * the most important in-viewport image should have `fetchpriority` set
+			 * to "high".
+			 */
+			if ( false === $maybe_in_viewport ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					esc_html__( 'An image should not be lazy-loaded and marked as high priority at the same time.', 'elementor' ),
+					''
+				);
+				/*
+				 * Set `fetchpriority` here for backward-compatibility as we should
+				 * not override what a developer decided, even though it seems
+				 * incorrect.
+				 */
+				$loading_attrs['fetchpriority'] = 'high';
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		if ( null === $maybe_in_viewport ) {
+			//@todo: Only run this functionality for frontend like in wp core.
+			$content_media_count = $this->increase_content_media_count( 0 );
+			$increase_count      = true;
+			if ( $content_media_count < $this->omit_threshold ) {
+				$maybe_in_viewport = true;
+			} else {
+				$maybe_in_viewport = false;
+			}
+		}
+
+		if ( $maybe_in_viewport ) {
+			$loading_attrs = $this->maybe_add_fetchpriority_high_attr( $loading_attrs, $attr );
+		} else {
+			$loading_attrs['loading'] = 'lazy';
+		}
+
+		if ( $increase_count ) {
+			$this->increase_content_media_count();
+		} elseif ( $maybe_increase_count ) {
+			if ( $this->min_priority_img_pixels <= $attr['width'] * $attr['height'] ) {
+				$this->increase_content_media_count();
+			}
+		}
+
+		return $loading_attrs;
+	}
+
+	/**
+	 * Keeps a count of media image.
+	 *
+	 * @param int $amount Amount by which count must be increased.
+	 * @return int current image count.
+	 */
+	private function increase_content_media_count( $amount = 1 ) {
+		static $content_media_count = 0;
+
+		$content_media_count += $amount;
+
+		return $content_media_count;
+	}
+
+	/**
+	 * Determines whether to add `fetchpriority='high'` to loading attributes.
+	 *
+	 * @param array  $loading_attrs Array of the loading optimization attributes for the element.
+	 * @param array  $attr          Array of the attributes for the element.
+	 * @return array Updated loading optimization attributes for the element.
+	 */
+	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ) {
+		if ( isset( $attr['fetchpriority'] ) ) {
+			if ( 'high' === $attr['fetchpriority'] ) {
```

Comment:
This one is a bit tricky, because we need to do two different things based on whether the attribute exists and if it has already been set to 'high' so any attributes added intentionally are respected and not overwritten here. That's why this is split into a nested set of conditionals.

Reply:
@joemcgill Is this not functionally equivalent? 🤔

```
if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
        $loading_attrs['fetchpriority'] = 'high';
        $this->high_priority_element_flag( false );
        return $loading_attrs;
}
```

Reply:
I see 👍